lib/kube/proxy/sess.go

```diff
@@ -1530,38 +1531,64 @@ func (s *session) retrieveEphemeralContainerCommand(ctx context.Context, usernam
 	return nil
 }
 
-// concurrentWaitGroup is a wait group that can be used concurrently.
-// It exits when the count reaches 0 for the first time.
-type concurrentWaitGroup struct {
-	cond  *sync.Cond
+// weakWaitGroup is a specialized synchronization primitive similar to sync.WaitGroup
+// but with **relaxed** guarantees. Unlike sync.WaitGroup, weakWaitGroup does not ensure
+// that the Wait() method will wait for all Add() calls to reach completion through Done()
+// if they are called concurrently. This means that there is a potential leak in the
+// synchronization of goroutines that are added to the weakWaitGroup and may be started
+// after the Wait() method is called.
+//
+// Use Case:
+// This weakWaitGroup is intended for scenarios where goroutines are initiated from
+// various parts of the codebase concurrently and need to be awaited only if they started before
+// a certain point in time, specifically before session.Close() is called. If a goroutine
+// is initiated after session.Close() has been invoked, it will not be included in the wait process.
+// It's the caller responsability to ensure that all goroutines started after Wait() returns end
+// up being a no-op.
+//
+// Important Considerations:
+//   - This implementation is UNSAFE as a general-purpose synchronization primitive.
+//   - It does not guarantee that Wait() will account for all Add() calls, leading to potential
+//     race conditions or goroutines that may not be properly awaited.
+//   - Due to these limitations, weakWaitGroup should be used with extreme caution and only
+//     in contexts where its relaxed guarantees are acceptable and safe.
+//
+// WARNING:
+// This is not a substitute for sync.WaitGroup in situations requiring strong synchronization
+// guarantees.
+type weakWaitGroup struct {
+	cond  sync.Cond
+	mu    sync.Mutex
 	count int
 }
 
-func newConcurrentWaitGroup() concurrentWaitGroup {
-	return concurrentWaitGroup{
-		cond:  sync.NewCond(&sync.Mutex{}),
-		count: 0,
-	}
-}
-
-func (c *concurrentWaitGroup) Add(delta int) {
-	c.cond.L.Lock()
-	defer c.cond.L.Unlock()
+func (c *weakWaitGroup) Add(delta int) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
 	c.count += delta
 }
 
-func (c *concurrentWaitGroup) Done() {
-	c.cond.L.Lock()
-	defer c.cond.L.Unlock()
+func (c *weakWaitGroup) Done() {
+	c.mu.Lock()
+	defer c.mu.Unlock()
 	c.count--
+	if c.cond.L == nil {
+		c.cond.L = &c.mu
+	}
```

Comment:
why? we can be absolutely certain that there's no waiters, if cond.L hasn't been set

Reply:
because the waiter is only called when the session is closed. It could happen that done is called and Wait wasn't yet called